Allow computation of VFRs on initial

GiudGiud opened the pull request:
The issue is only with Rhie Chow interpolation (average is fine)

- we need RC UO before VFRs on initial otherwise we hit a missing a functor
- by default, general UOs are after side UOs, this cannot be changed, many tests depend on it
- preaux for RC UO is not great if using auxvariables for advection, let's have a warning
- preic is even worse because the momentum advection kernels are not initialized

Other solutions:
- initialize a data everywhere on flow domain with 0s and return average interpolation on INITIAL

- [x] Add tests
- [x] Add an option in the action to get it right by default maybe

GiudGiud committed:
Give user a workflow for running volumetric flow rates on initial:
- enable the execution of RC UO on other flags
- error if VFR comes before RC UO
- warning if RC UO is used on preaux with auxvariables
closes #18817

GiudGiud committed:
Misc comments in the routine triggering the execution of user objects

GiudGiud committed:
Add test for new warnings and errors around execution at initial, refs #18817

GiudGiud committed:
Make action execute RC UO on initial by default, refs #18817

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1187851/) on 863622f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22245/site)

This comment will be updated on new commits.

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -45,7 +47,9 @@ INSFVRhieChowInterpolator::validParams()
   ExecFlagEnum & exec_enum = params.set<ExecFlagEnum>("execute_on", true);
   exec_enum.addAvailableFlags(EXEC_PRE_KERNELS);
   exec_enum = {EXEC_PRE_KERNELS};
```
It looks like you're setting the default to include initial within the action. If you want that, let's just make it the default here so that users running with the action and with a full input file get the same default behavior

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -202,6 +206,11 @@ INSFVRhieChowInterpolator::INSFVRhieChowInterpolator(const InputParameters & par
   if (_velocity_interp_method != Moose::FV::InterpMethod::Average)
     fillARead();
+
+  if (!getExecuteOnEnum().contains(EXEC_PRE_KERNELS) || getExecuteOnEnum().size() > 2 ||
```
I feel like we should error if it doesn't contain `EXEC_PRE_KERNELS` ... but I suppose if you're doing passive scalar transport (or similar) in an app different from the Navier-Stokes solve, and so are reading the A coefficients from aux variables, then you don't actually need `EXEC_PRE_KERNELS` do you?

GiudGiud replied:
well the execution is very cheap since we wont have any residual objects if we do passive scalar transport. 

I think you are right, I should just make EXEC_INITIAL default and we ll take the small hit on performance to avoid the trouble.

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -219,6 +228,9 @@ INSFVRhieChowInterpolator::fillARead()
 
     _a_data_provided = true;
     _a_aux.resize(libMesh::n_threads());
+    if (getParam<bool>("force_preaux"))
+      mooseWarning("Executing the Rhie Chow user object before auxkernels is only valid if the "
+                   "auxvariables involved are constant over linear iterations");
```
It seems like we will emit warnings when using the action even when the user themselves have not requested any non-standard behavior. This can train them to not respect warnings ... but at the same time it is good to tell them where possible issues might arise. 

Yucko. This is a really good example of wanting the arbitrary dep resolution. Just having it for user objects alone would fix this problem.

GiudGiud replied:
true true. Maybe that is the smaller scope to carve out on this

lindsayad replied:
I've created #22255 which I don't think should be very hard to implement

GiudGiud replied:
ok so now we are hitting errors with the PINSFV + reconstruction because the initial setup on a parsed function is not being run before preaux I think ...
lovely haha

lindsayad replied:
pre-aux user object `execute` should be run after functions have their `initialSetup` done

GiudGiud replied:
This would make sense but smh getting a null ptr at evaluate time. 
I can investigate. 

Is the other modification OK ? (the flag on area PP)

lindsayad reviewed (changes requested):
I'm guessing you tried swapping the execution location of general user objects?

GiudGiud commented:
Yes I tried re-ordering the user objects. 
It did not go well :S

GiudGiud committed:
Make behavior of RC UO consistent between action and regular
input file syntax
irefs #18817

GiudGiud committed:
Make sure we never de-reference an uninitialized parsed function

Gather surface areas as early as possible so that they are available for other UOs
done for #18817

GiudGiud commented:
ok see my two last commits for the dumb stuff I have had to do to make this work.

If the cost to the framework is too high, I m happy to disallow VFR pps on INITIAL for finite volume with Rhie Chow interpolation.
There is no way around the dependency resolution issues without the general dependency resolution framework.
I have now gained significant motivation to work on that. Maybe once I m done with my other tasks for the FY I could go full time on it

GiudGiud commented:
the hack with the parsed function would be also necessary for these functions since they rely on an initialSetup after their constructor to be ready for use
```
SolutionFunction.C:SolutionFunction::initialSetup()
ImageFunction.C:ImageFunction::initialSetup()
Axisymmetric2D3DSolutionFunction.C:Axisymmetric2D3DSolutionFunction::initialSetup()
PiecewiseLinearBase.C:PiecewiseLinearBaseTempl<BaseClass>::initialSetup()
MooseParsedVectorFunction.C:MooseParsedVectorFunction::initialSetup()
PiecewiseConstantFromCSV.C:PiecewiseConstantFromCSV::initialSetup()
MooseParsedGradFunction.C:MooseParsedGradFunction::initialSetup()
```

lindsayad commented on framework/src/functions/MooseParsedFunction.C:
```diff
@@ -43,20 +43,26 @@ template <typename T>
 Real
 MooseParsedFunctionTempl<T>::value(Real t, const Point & p) const
 {
+  if (!_function_ptr)
+    const_cast<MooseParsedFunctionTempl<T> *>(this)->initialSetup();
```
I don't think this can fly, or at least I'd want to see all other avenues exhausted before we went with it

GiudGiud replied:
ok removed. This was indeed the worst, but the intention was pure so I added an assert

lindsayad commented on framework/src/postprocessors/AreaPostprocessor.C:
```diff
@@ -18,6 +18,9 @@ AreaPostprocessor::validParams()
 
   params.addClassDescription("Computes the \"area\" or dimension - 1 \"volume\" of a given "
                              "boundary or boundaries in your mesh.");
+  // Other user objects may depend on this value on INITIAL
+  // pre-initial conditions is as early as possible
+  params.set<bool>("force_preic") = true;
```
I think this is ok, but I'd like to understand why it's necessary?

GiudGiud replied:
because you run the flux BCs using the area pps on INITIAL because the RC UO is ran on INITIAL. 

We dont really need to run the flux BCs since the issue with VFRs is inside the domain, but detecting this on INITIAL and changing the list of residual object was harder

lindsayad replied:
Shouldn't pre-aux be sufficient? E.g. run at the "same time" as the RC UO? Because side user objects are run before general user objects

GiudGiud replied:
yeah could be. I ll give it a try

GiudGiud commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -20,7 +20,9 @@
 #include "INSFVPressureVariable.h"
 #include "PiecewiseByBlockLambdaFunctor.h"
 #include "VectorCompositeFunctor.h"
+#include "FVKernel.h"
```
```suggestion
```

GiudGiud commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -20,7 +20,9 @@
 #include "VectorCompositeFunctor.h"
+#include "FVKernel.h"
 #include "FVElementalKernel.h"
+#include "FVFluxBC.h"
```
```suggestion
```

GiudGiud reviewed (commented):
I ll merge in those two comments with yours if you have any @lindsayad

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -263,6 +270,11 @@ INSFVRhieChowInterpolator::fillARead()
 void
 INSFVRhieChowInterpolator::initialSetup()
 {
+  if (!_initial_setup_done)
+    insfvSetup();
```
```suggestion
  insfvSetup();
```

lindsayad commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -263,6 +270,11 @@ INSFVRhieChowInterpolator::fillARead()
+  if (!_initial_setup_done)
+    insfvSetup();
+
+  _initial_setup_done = true;
+
```
```suggestion
```

lindsayad commented on modules/navier_stokes/src/userobjects/PINSFVRhieChowInterpolator.C:
```diff
@@ -74,6 +74,19 @@ PINSFVRhieChowInterpolator::PINSFVRhieChowInterpolator(const InputParameters & p
+void
+PINSFVRhieChowInterpolator::initialSetup()
+{
+  if (!_initial_setup_done)
+  {
+    insfvSetup();
+    pinsfvSetup();
+  }
+
+  _initial_setup_done = true;
+  INSFVRhieChowInterpolator::initialSetup();
```
```suggestion
  INSFVRhieChowInterpolator::initialSetup();
  pinsfvSetup();
```

lindsayad reviewed (changes requested):
I think you should be able to remove the `_initial_setup_done` flag and also the `residualSetup` overrides in both `INSFVRhieChowInterpolator` and `PINSFVRhieChowInterpolator`

GiudGiud committed:
Move residual setup to initial setup in RC interpolator for free flow
refs  #18817

GiudGiud committed:
Execute porosity smoothing right before execution instead of in residualSetup:
- later than initialSetup so functions are already initialized (Functions get initialized after UOs)
- before execution on INITIAL (residualSetup was after INIITAL)
refs #18817

GiudGiud committed:
Avoid time kernels when executing RC on initial as they are not initialized
refs  #18817

GiudGiud committed:
Loosen tolerance on a FE test of flow enar a wedge
Tighten tolerance of solve on a FV test
Both had small diffs at various distributed mesh number of processes

GiudGiud committed:
Apply suggestions from code review

Apply suggestions from code review

Clarify comment refs #18817

GiudGiud commented on modules/navier_stokes/test/tests/postprocessors/flow_rates/tests:
```diff
@@ -131,4 +140,4 @@
     []
   []
-[]
+[]
```
```suggestion
[]

```

GiudGiud commented:
I ll squash those commits when addressing the reviews

lindsayad commented on framework/src/postprocessors/AreaPostprocessor.C:
```diff
@@ -18,6 +18,9 @@ AreaPostprocessor::validParams()
   params.addClassDescription("Computes the \"area\" or dimension - 1 \"volume\" of a given "
                              "boundary or boundaries in your mesh.");
+  // Other user objects may depend on this value on INITIAL
+  // preaux runs before most user objects, and side UOs run before all UOs
```
This comment should be clarified. Elemental user objects *execute* before side user objects on a per-element basis, but side user objects are joined and finalized before element user objects

GiudGiud replied:
better now?
this is actually:
GeneralUO needs BC computeRC needs SideUO(area)

lindsayad reviewed (changes requested):
In the middle of something ... will finish review soon

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1193897/) on 11283bb wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>be5a42</th>
      <th colspan=4>#22245 11283b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.29%</td>
      <td>84.29%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80089</td>
      <td>80091</td>
      <td>+2</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14928</td>
      <td>14927</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22245/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22245/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>be5a42</th>
      <th colspan=4>#22245 11283b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.50%</td>
      <td>85.50%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>9765</td>
      <td>9768</td>
      <td>+3</td>
      <td>15</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1656</td>
      <td>1656</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22245/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22245/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22245/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on modules/navier_stokes/src/loops/GatherRCDataElementThread.C:
```diff
@@ -48,7 +49,14 @@ GatherRCDataElementThread::subdomainChanged()
     for (auto var_ek : var_eks)
       if (auto insfv_ek = dynamic_cast<INSFVMomentumResidualObject *>(var_ek))
+      {
+        // On INITIAL, the time integrator is not ready for these residual objects
+        if (this->_fe_problem.getCurrentExecuteOnFlag() == EXEC_INITIAL)
+          if (dynamic_cast<FVTimeKernel *>(var_ek))
+            continue;
```
This is where I wonder what the value of executing the VFR on initial is. This will lead to computation of a different velocity than what is used in the first residual evaluation

GiudGiud replied:
It wont be that different and it will only matter for internal sidesets.
VFR on external boundaries on initial seems important. 

Do we have a way to know if a sideset is internal or on the boundary of the flow domain? Since this issue is for RC coefficients, it's a problem inside the flow domain not on its edge. We could disallow only inside

lindsayad commented:
I am fine with the change to how PINSFV is setup BTW

GiudGiud committed:
Revert execution of RC Interpolator on INITIAL because we could not obtain the same RC coefficients because we could not compute the time derivatives because the time integrator was not ready.

Reverting entailed:
- forbid execution on initial of volumetric flow rates (VFR) on internal boundaries
- regold test that involved internal boundary VFRs
- add an exception check for internal boundary VFRs
- add a routine to know whether a boundary is internal to a group of blocks
- re-adjust setup of PINSFV interpolator to avoid setting up on execute as this was less clean. Setting up on initialSetup is still not possible, this is documented in the VFR and the UO
Closes #18817

GiudGiud committed:
Add -debug to git ignore for people typing METHOD=debug make instead of METHOD=dbg
same for equivalents for opt and profiling

GiudGiud reviewed (commented):
@lindsayad ready for review again

lindsayad commented on .gitignore:
```diff
@@ -6,6 +6,7 @@
 *.a
 *-opt
 *-dbg
+*-debug
```
where does this come from?

GiudGiud replied:
it s missing. You could be checking in your debug executable by accident currently

sorry for the rider commit, I can boot it to a fixup PR if you prefer

lindsayad replied:
the debug executable is `dbg` which is already in the `.gitignore`

lindsayad replied:
```suggestion
```

GiudGiud replied:
You can compile with METHOD=debug and it works too
``
(moose) giudgl@fn654105:~/projects/moose_v2/framework$ ls
Makefile               build_support          include                libmoose-devel.0.dylib notes.md
app.mk                 contrib                libmoose-debug.0.dylib libmoose-devel.dylib   plugin.mk
build                  data                   libmoose-debug.dylib   libmoose-devel.la      scripts
build.mk               doc                    libmoose-debug.la      moose.mk               src
(moose) giudgl@fn654105:~/projects/moose_v2/framework$ ldd libmoose-debug.dylib 
libmoose-debug.dylib:
	/Users/giudgl/projects/moose_v2/framework/libmoose-debug.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/giudgl/projects/moose_v2/framework/contrib/pcre/libpcre-debug.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/../libmesh/lib/libmesh_dbg.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/../libmesh/lib/libtimpi_dbg.11.dylib (compatibility version 12.0.0, current version 12.0.0)
	@rpath/libhdf5_cpp.200.dylib (compatibility version 202.0.0, current version 202.0.0)
	@rpath/libvtkIOCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkCommonCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkCommonDataModel-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkFiltersCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkIOXML-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkImagingCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkIOImage-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkImagingMath-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkIOParallelXML-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkParallelMPI-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkParallelCore-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libvtkCommonExecutionModel-9.1.1.dylib (compatibility version 1.0.0, current version 9.1.0)
	@rpath/libz.1.dylib (compatibility version 1.0.0, current version 1.2.12)
	@rpath/libslepc.3.16.dylib (compatibility version 3.16.0, current version 3.16.2)
	@rpath/libpetsc.3.16.dylib (compatibility version 3.16.0, current version 3.16.6)
	@rpath/libHYPRE-2.23.0.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libstrumpack.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libsuperlu_dist.7.dylib (compatibility version 7.0.0, current version 7.1.1)
	@rpath/libhdf5_hl.200.dylib (compatibility version 201.0.0, current version 201.1.0)
	@rpath/libhdf5.200.dylib (compatibility version 202.0.0, current version 202.0.0)
	@rpath/libparmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmetis.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmpifort.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libpmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libgfortran.5.dylib (compatibility version 6.0.0, current version 6.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.100.3)
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libmpi.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libmpicxx.12.dylib (compatibility version 15.0.0, current version 15.2.0)
	@rpath/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
``

GiudGiud replied:
```
(moose) giudgl@fn654105:~/projects/moose_v2/modules/navier_stokes$ ll
total 672
drwxr-xr-x  17 giudgl  INEL-NT\Domain Users     544 Oct 13 17:59 ./
drwxr-xr-x  34 giudgl  INEL-NT\Domain Users    1088 Oct 12 19:06 ../
drwxr-xr-x   2 giudgl  INEL-NT\Domain Users      64 Oct 13 17:59 .libs/
-rw-r--r--   1 giudgl  INEL-NT\Domain Users  198997 Oct 13 17:59 .previous_test_results.json
-rw-r--r--   1 giudgl  INEL-NT\Domain Users    1077 Sep 20 08:30 Makefile
drwxr-xr-x   4 giudgl  INEL-NT\Domain Users     128 Oct 13 17:58 build/
drwxr-xr-x  10 giudgl  INEL-NT\Domain Users     320 Sep 27 22:51 doc/
drwxr-xr-x   4 giudgl  INEL-NT\Domain Users     128 Sep 27 22:51 examples/
drwxr-xr-x  18 giudgl  INEL-NT\Domain Users     576 Sep 27 22:51 include/
drwxr-xr-x   6 giudgl  INEL-NT\Domain Users     192 Oct 13 17:59 lib/
-rwxr-xr-x   1 giudgl  INEL-NT\Domain Users  132960 Oct 13 17:59 navier_stokes-debug*
drwxr-xr-x   3 giudgl  INEL-NT\Domain Users      96 Sep 20 08:30 python/
lrwxr-xr-x   1 giudgl  INEL-NT\Domain Users      23 Sep 20 08:30 run_tests@ -> ../../scripts/run_tests
drwxr-xr-x  22 giudgl  INEL-NT\Domain Users     704 Oct 13 17:59 src/
drwxr-xr-x   6 giudgl  INEL-NT\Domain Users     192 Oct 13 17:59 test/
-rw-r--r--   1 giudgl  INEL-NT\Domain Users      92 Sep 20 08:30 testroot
drwxr-xr-x   5 giudgl  INEL-NT\Domain Users     160 Sep 27 22:51 unit/
(moose) giudgl@fn654105:~/projects/moose_v2/modules/navier_stokes$ METHOD=debug ./run_tests 
test:finite_element/ins/RZ_cone.jac ....................................................... [METHOD!=OPT] SKIP
test:finite_element/ins/RZ_cone.ad_rz_cone_by_parts_traction_steady_supg_pspg_jac ......... [METHOD!=OPT] SKIP
test:finite_element/ins/RZ_cone.ad_jac .................................................... [METHOD!=OPT] SKIP
test:finite_volume/ins/channel-flow.triangles/tets .............................................. [HEAVY] SKIP
test:finite_volume/ins/channel-flow.mixing-length-jac ..................................... [METHOD!=OPT] SKIP
test:finite_element/ins/lid_driven.ad_lid_driven_action ................ [METHOD!=OPT,skipped dependency] SKIP
test:finite_element/ins/lid_driven.ad_lid_driven_mean_zero_pressure ....................... [METHOD!=OPT] SKIP
test:finite_element/ins/lid_driven.still_unstable ......................................... [METHOD!=OPT] SKIP
test:finite_element/ins/lid_driven.ad_lid_driven .......................................... [METHOD!=OPT] SKIP
test:finite_element/ins/lid_driven.ad_lid_driven_stabilized_jacobian ...................... [METHOD!=OPT] SKIP
test:finite_element/ins/lid_driven.ad_lid_driven_jacobian ................................. [METHOD!=OPT] SKIP
test:finite_element/ins/coupled-force.steady-function-jac ................................. [METHOD!=OPT] SKIP
test:finite_element/ins/coupled-force.steady-action-jac ................................... [METHOD!=OPT] SKIP
test:finite_element/ins/coupled-force.steady-action-function-jac .......................... [METHOD!=OPT] SKIP
test:finite_element/ins/coupled-force.steady-jac .......................................... [METHOD!=OPT] SKIP
test:finite_element/ins/mms/supg.alpha_1e-3/4x4 ................................................. [HEAVY] SKIP
test:finite_element/ins/mms/supg.alpha_1e-6/8x8 ................................................. [HEAVY] SKIP
test:finite_element/ins/mms/supg.alpha_1e-3/16x16 ............................................... [HEAVY] SKIP
test:finite_element/ins/mms/supg.alpha_1e-6/32x32 ............................................... [HEAVY] SKIP
test:finite_element/ins/mms/supg.alpha_1e-6/4x4 ................................................. [HEAVY] SKIP
test:finite_element/ins/mms/supg.alpha_1e-3/8x8 ................................................. [HEAVY] SKIP
test:finite_element/ins/mms/supg.alpha_1e-6/16x16 ............................................... [HEAVY] SKIP
test:finite_element/ins/mms/supg.alpha_1e-3/32x32 ............................................... [HEAVY] SKIP
test:finite_element/ins/mms/pspg.alpha_1e-3/4x4 ................................................. [HEAVY] SKIP
test:finite_element/ins/mms/pspg.alpha_1e-6/8x8 ................................................. [HEAVY] SKIP
test:finite_element/ins/mms/pspg.alpha_1e-3/32x32 ............................................... [HEAVY] SKIP
test:finite_element/ins/mms/pspg.alpha_1e-6/32x32 ............................................... [HEAVY] SKIP
test:finite_element/ins/mms/pspg.alpha_1e-3/16x16 ............................................... [HEAVY] SKIP
test:finite_element/ins/mms/pspg.alpha_1e-6/4x4 ................................................. [HEAVY] SKIP
test:finite_element/ins/mms/pspg.alpha_1e-3/8x8 ................................................. [HEAVY] SKIP
test:finite_element/ins/mms/pspg.alpha_1e-6/16x16 ............................................... [HEAVY] SKIP
test:finite_volume/ins/lid-driven.jacobian ................................................ [METHOD!=OPT] SKIP
test:finite_volume/ins/lid-driven.quiescent_jac ........................................... [METHOD!=OPT] SKIP
test:finite_volume/ins/mms/channel-flow.plane-poiseuille-rc-first ............................... [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow.plane-poiseuille-average-first .......................... [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow.2d-rc ................................................... [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow.2d-rc-with-temp ......................................... [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow.plane-poiseuille-rc ..................................... [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow.2d-average-with-temp .................................... [HEAVY] SKIP
test:finite_element/ins/boussinesq.jac .................................................... [METHOD!=OPT] SKIP
test:finite_volume/wcns/channel-flow.transient-jac ........................................ [METHOD!=OPT] SKIP
test:finite_volume/wcns/channel-flow.turbulence-jac ....................................... [METHOD!=OPT] SKIP
test:finite_volume/pins/mms/porosity_change.pressure-corrected-jac ........................ [METHOD!=OPT] SKIP
test:finite_element/ins/boussinesq/benchmark.1e5 ................................................ [HEAVY] SKIP
test:finite_element/ins/boussinesq/benchmark.1e6 ................................................ [HEAVY] SKIP
test:finite_element/ins/boussinesq/benchmark.1e4 ................................................ [HEAVY] SKIP
test:finite_element/ins/boussinesq/benchmark.1e3 ................................................ [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow/cylindrical.rc .......................................... [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow/cylindrical.rc-with-temp ................................ [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow/rotated.plane-poiseuille-average ........................ [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow/rotated.plane-poiseuille-average-first .................. [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow/rotated.plane-poiseuille-rc ............................. [HEAVY] SKIP
test:finite_volume/ins/mms/channel-flow/rotated.plane-poiseuille-rc-first ....................... [HEAVY] SKIP
test:finite_volume/pwcns/channel-flow.transient-jac ....................................... [METHOD!=OPT] SKIP
test:finite_volume/ins/mms/skew-correction.vortex-skewness-corrected-action ..................... [HEAVY] SKIP
test:finite_volume/ins/mms/skew-correction.vortex-skewness-corrected ............................ [HEAVY] SKIP
test:finite_volume/ins/mixing_length_total_viscosity_material.capped_mixing_length/transient .... [HEAVY] SKIP
test:finite_element/ins/lid_driven/mixed-transient-steady.jac ............................. [METHOD!=OPT] SKIP
test:finite_volume/ins/mixing_length_eddy_viscosity_aux.capped_mixing_length/transient .......... [HEAVY] SKIP
test:finite_volume/ins/wall_distance_capped_mixing_length_aux.capped_mixing_length .............. [HEAVY] SKIP
test:finite_volume/ins/mms/cylindrical.2d-rc-diri ............................................... [HEAVY] SKIP
test:finite_volume/ins/action/errors.momentum-inlet-error ................................................. OK
```

lindsayad replied:
Haha you are right. So if we are doing those extensions, and I agree we should, you might as well add all of them
```suggestion
*-debug
*-optimized
*-profiling
*-pro
*-oprofile
```

GiudGiud replied:
Ok Ill amend the commit and re-push

GiudGiud replied:
done

lindsayad commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1054,6 +1054,44 @@ MooseMesh::getBoundaryActiveNeighborElemIds(BoundaryID bid) const
+                                          const std::set<SubdomainID> & blk_group) const
+{
+  mooseAssert(_bnd_elem_range, "Boundary element range is not initialized");
+  const bool all_blocks =
+      std::find(blk_group.begin(), blk_group.end(), Moose::ANY_BLOCK_ID) != blk_group.end();
```
```suggestion
  const bool all_blocks = blk_group.find(Moose::ANY_BLOCK_ID) != blk_group.end();
```
This is O(log(N)) as opposed to `std::find` which is O(N)

lindsayad commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1054,6 +1054,44 @@ MooseMesh::getBoundaryActiveNeighborElemIds(BoundaryID bid) const
+        // then the boundary cuts the subdomains
+        if (all_blocks || std::find(blk_group.begin(), blk_group.end(), neighbor->subdomain_id()) !=
+                              blk_group.end())
+          return true;
+      }
+    }
+  }
+  return false;
```
It almost feels like this logic is the opposite of what you want. It seems like if you ever encounter a case in which you don't have a neighbor or either the element or neighbor subdomains don't match the subdomains in the provided set that you should immediately return `false`. And then only if you examine all the triples for which the boundary ID matches the provided `bid` and they meet the internal criteria would you return `true`

lindsayad replied:
I think the concept in this method is appropriate if you name this method something like `isAnyPartOfBoundaryInternalToSubdomains`

GiudGiud replied:
Let's reverse it and say external

lindsayad commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1054,6 +1054,44 @@ MooseMesh::getBoundaryActiveNeighborElemIds(BoundaryID bid) const
+    if (elem_bid == bid)
+    {
+      // If an element is internal to the group of subdomain, check the neighbor
+      if (all_blocks || std::find(blk_group.begin(), blk_group.end(), elem_ptr->subdomain_id()) !=
+                            blk_group.end())
```
```suggestion
      if (all_blocks || blk_group.find(elem_ptr->subdomain_id()) != blk_group.end())
```

lindsayad commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1054,6 +1054,44 @@ MooseMesh::getBoundaryActiveNeighborElemIds(BoundaryID bid) const
+          continue;
+        // If the neighbor is also in the group of subdomain,
+        // then the boundary cuts the subdomains
+        if (all_blocks || std::find(blk_group.begin(), blk_group.end(), neighbor->subdomain_id()) !=
+                              blk_group.end())
```
```suggestion
        if (all_blocks || blk_group.find(neighbor->subdomain_id()) != blk_group.end())
```

lindsayad commented on modules/navier_stokes/src/loops/GatherRCDataElementThread.C:
```diff
@@ -48,7 +49,14 @@ GatherRCDataElementThread::subdomainChanged()
     for (auto var_ek : var_eks)
       if (auto insfv_ek = dynamic_cast<INSFVMomentumResidualObject *>(var_ek))
+      {
+        // On INITIAL, the time integrator is not ready for these residual objects
```
So... if you are not executing VFRs on internal boundaries, then can't you simply not execute the RC user object on initial? E.g. can we just keep carrying on with the RC UO only executing on `EXEC_PRE_KERNELS`? Then we don't have to do this stuff here

GiudGiud replied:
oh I forgot to delete that code. Good catch

I removed the EXEC_INITIAL from the RC UO already

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1054,6 +1054,44 @@ MooseMesh::getBoundaryActiveNeighborElemIds(BoundaryID bid) const
 }
 
+bool
+MooseMesh::isBoundaryInternalToSubdomains(BoundaryID bid,
```
```suggestion
MooseMesh::isBoundaryFullyExternalToSubdomains(BoundaryID bid,
```

GiudGiud commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -447,6 +447,16 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+   * @param blk_group the group of blocks potentially traversed
+   * @return whether the boundary crosses between the subdomains in the group
+   */
+  bool isBoundaryInternalToSubdomains(BoundaryID bid,
```
```suggestion
  bool isBoundaryFullyExternalToSubdomains(BoundaryID bid,
```

lindsayad replied:
I like that name

GiudGiud commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -447,6 +447,16 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
   std::unordered_set<dof_id_type> getBoundaryActiveNeighborElemIds(BoundaryID bid) const;
 
+  /**
+   * Returns whether a boundary (given by its id) is crossing through a group of blocks,
```
```suggestion
   * Returns whether a boundary (given by its id) is not crossing through a group of blocks,
```

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1054,6 +1054,44 @@ MooseMesh::getBoundaryActiveNeighborElemIds(BoundaryID bid) const
+        // then the boundary cuts the subdomains
+        if (all_blocks || std::find(blk_group.begin(), blk_group.end(), neighbor->subdomain_id()) !=
+                              blk_group.end())
+          return true;
```
```suggestion
          return false;
```

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -1054,6 +1054,44 @@ MooseMesh::getBoundaryActiveNeighborElemIds(BoundaryID bid) const
+      }
+    }
+  }
+  return false;
```
```suggestion
  return true;
```

GiudGiud commented on framework/src/postprocessors/AreaPostprocessor.C:
```diff
@@ -18,6 +18,10 @@ AreaPostprocessor::validParams()
 
   params.addClassDescription("Computes the \"area\" or dimension - 1 \"volume\" of a given "
                              "boundary or boundaries in your mesh.");
+  // Other user objects may depend on this value on INITIAL
+  // preaux runs will make it execute before most types of user objects. Changed for
+  // NavierStokes: this is needed by flux boundary conditions, executed by a general user object
+  params.set<bool>("force_preaux") = true;
```
```suggestion
```

GiudGiud commented on modules/navier_stokes/src/actions/NSFVAction.C:
```diff
@@ -2228,7 +2228,9 @@ NSFVAction::addBoundaryPostprocessors()
       const std::string pp_type = "AreaPostprocessor";
       InputParameters params = _factory.getValidParams(pp_type);
       params.set<std::vector<BoundaryName>>("boundary") = {_inlet_boundaries[bc_ind]};
+      // Get execution before Rhie Chow user object on INITIAL
```
```suggestion
```

GiudGiud commented on modules/navier_stokes/src/actions/NSFVAction.C:
```diff
@@ -2228,7 +2228,9 @@ NSFVAction::addBoundaryPostprocessors()
       params.set<std::vector<BoundaryName>>("boundary") = {_inlet_boundaries[bc_ind]};
+      // Get execution before Rhie Chow user object on INITIAL
       params.set<ExecFlagEnum>("execute_on") = EXEC_INITIAL;
+      params.set<bool>("force_preaux") = true;
```
```suggestion
```

GiudGiud commented on modules/navier_stokes/src/loops/GatherRCDataElementThread.C:
```diff
@@ -48,7 +49,14 @@ GatherRCDataElementThread::subdomainChanged()
     for (auto var_ek : var_eks)
       if (auto insfv_ek = dynamic_cast<INSFVMomentumResidualObject *>(var_ek))
+      {
+        // On INITIAL, the time integrator is not ready for these residual objects
+        if (this->_fe_problem.getCurrentExecuteOnFlag() == EXEC_INITIAL)
+          if (dynamic_cast<FVTimeKernel *>(var_ek))
+            continue;
```
```suggestion
```

GiudGiud committed:
Apply suggestions from code review:
- rename routine to introduce notion that it has to be fully external (or equivalently not entirely internal)
- remove more accomodations for executing RC on initial, which are not necessary
- use vector.find instead of std::find(vector) as it is faster

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1201590/) on eea7875 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22245/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22245/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 43cb5b773f57176eda58683ccd3d00513b05f1de`

lindsayad commented on modules/navier_stokes/src/loops/GatherRCDataElementThread.C:
```diff
@@ -48,7 +49,10 @@ GatherRCDataElementThread::subdomainChanged()
     copied_queries.template condition<AttribVar>(static_cast<int>(var_num)).queryInto(var_eks);
     for (auto var_ek : var_eks)
       if (auto insfv_ek = dynamic_cast<INSFVMomentumResidualObject *>(var_ek))
+      {
+
         _insfv_elemental_kernels.push_back(insfv_ek);
+      }
```
```suggestion
        _insfv_elemental_kernels.push_back(insfv_ek);
```

GiudGiud replied:
done

GiudGiud commented:
Thanks for the review!
And thanks to @muelcm for bringing it to our attention
